cpp/src/arrow/util/iterator.h

```diff
@@ -312,4 +314,128 @@ Iterator<T> MakeFlattenIterator(Iterator<Iterator<T>> it) {
   return Iterator<T>(FlattenIterator<T>(std::move(it)));
 }
 
+namespace detail {
+
+// A type-erased promise object for ReadaheadQueue.
+struct ARROW_EXPORT ReadaheadPromise {
+  virtual ~ReadaheadPromise();
+  virtual void Call() = 0;
+};
+
+template <typename T>
+struct ReadaheadIteratorPromise : ReadaheadPromise {
+  ~ReadaheadIteratorPromise() override {}
+
+  explicit ReadaheadIteratorPromise(Iterator<T>* it) : it_(it) {}
+
+  void Call() override {
+    assert(!called_);
+    status_ = it_->Next(&out_);
+    called_ = true;
+  }
+
+  Iterator<T>* it_;
+  Status status_;
+  T out_ = IterationTraits<T>::End();
+  bool called_ = false;
+};
+
+class ARROW_EXPORT ReadaheadQueue {
+ public:
+  explicit ReadaheadQueue(int readahead_queue_size);
+  ~ReadaheadQueue();
+
+  Status Append(std::unique_ptr<ReadaheadPromise>);
+  Status PopDone(std::unique_ptr<ReadaheadPromise>*);
+  Status Pump(std::function<std::unique_ptr<ReadaheadPromise>()> factory);
+  Status Shutdown();
+  void EnsureShutdownOrDie();
+
+ protected:
+  class Impl;
+  std::shared_ptr<Impl> impl_;
+};
+
+}  // namespace detail
+
+/// \brief Readahead iterator that iterates on the underlying iterator in a
+/// separate thread, getting up to N values in advance.
+template <typename T>
+class ReadaheadIterator {
+  using PromiseType = typename detail::ReadaheadIteratorPromise<T>;
+
+ public:
+  // Public default constructor creates an empty iterator
+  ReadaheadIterator() : done_(true) {}
+
+  ~ReadaheadIterator() {
+    if (queue_) {
+      // Make sure the queue doesn't call any promises after this object
+      // is destroyed.
+      queue_->EnsureShutdownOrDie();
+    }
+  }
+
+  ARROW_DEFAULT_MOVE_AND_ASSIGN(ReadaheadIterator);
+  ARROW_DISALLOW_COPY_AND_ASSIGN(ReadaheadIterator);
+
+  Status Next(T* out) {
+    if (done_) {
+      *out = IterationTraits<T>::End();
+      return Status::OK();
+    }
+
+    std::unique_ptr<detail::ReadaheadPromise> promise;
+    ARROW_RETURN_NOT_OK(queue_->PopDone(&promise));
+    auto it_promise = static_cast<PromiseType*>(promise.get());
+
+    ARROW_RETURN_NOT_OK(queue_->Append(MakePromise()));
+
+    ARROW_RETURN_NOT_OK(it_promise->status_);
+    *out = std::move(it_promise->out_);
+    if (*out == IterationTraits<T>::End()) {
+      done_ = true;
+      // Shutdown the queue, but it may still have requested a couple more values
+      // in the meantime.
+      ARROW_RETURN_NOT_OK(queue_->Shutdown());
```

Comment:
Note to self: no need to actually shutdown the queue (`Shutdown(wait=true)` may take some time to return, and `Shutdown(wait=false)` is unsafe here).